src/org/jetbrains/java/decompiler/modules/decompiler/SwitchPatternMatchProcessor.java

```diff
@@ -48,60 +47,173 @@ private static boolean processStatement(SwitchStatement stat, Statement root) {
+                  }
+                  guards.put(stat.getCaseValues().get(i), guardExprent);
+                  // eliminate the guard `if`, alongside the assignment if not inverted
+                  guardIf.replaceWithEmpty();
+                  guardIf.getParent().getStats().remove(0);
+                  Statement nextStat = guardIf.getParent().getStats().get(0);
+                  // add the pattern variable assignment (or case code for inverted cases) to next statement
+                  if (nextStat instanceof BasicBlockStatement) {
+                    nextStat.getExprents().addAll(0, castExprent);
+                  } else {
+                    nextStat.getFirst().getExprents().addAll(0, castExprent);
+                  }
+                  break;
+                }
+              }
+            }
+          }
+        }
+      }
     }
 
     for (int i = 0; i < stat.getCaseStatements().size(); i++) {
       Statement caseStat = stat.getCaseStatements().get(i);
 
-      if (!(caseStat instanceof BasicBlockStatement)) {
-        continue;
-      }
-
-      Exprent caseExpr = stat.getCaseValues().get(i).get(0);
+      List<Exprent> allCases = stat.getCaseValues().get(i);
+      Exprent caseExpr = allCases.get(0);
 
-      // Default branch
+      // null expression = default case, can't be shared with patterns
       if (caseExpr == null) {
         continue;
       }
 
+      if (guards.containsKey(allCases)) {
+        // add the guard to the same index as this case, padding the list with nulls as necessary
+        while(stat.getCaseGuards().size() <= i)
+          stat.getCaseGuards().add(null);
+        stat.getCaseGuards().set(i, guards.get(allCases));
+      }
       if (caseExpr instanceof ConstExprent) {
         int caseValue = ((ConstExprent)caseExpr).getIntValue();
 
+        // -1 always means null
         if (caseValue == -1) {
-          // null
-          stat.getCaseValues().get(i).set(0, new ConstExprent(VarType.VARTYPE_NULL, null, null));
+          allCases.remove(caseExpr);
+          ConstExprent nullConst = new ConstExprent(VarType.VARTYPE_NULL, null, null);
+          // null can be shared with a pattern or default; put it at the end, but before default, to make sure it doesn't get
+          // absorbed by the default or overwritten by a pattern
+          if (allCases.contains(null)) {
+            allCases.add(allCases.indexOf(null), nullConst);
+          } else {
+            allCases.add(nullConst);
+          }
         }
       }
 
-      // Make instanceof
+      // find the pattern variable assignment
+      if (caseStat instanceof SequenceStatement) {
+        Statement oldStat = caseStat;
+        caseStat = caseStat.getStats().get(0);
+        // we can end up with a SequenceStatement with 1 statement from guard `if` elimination, eliminate the sequence entirely
+        if (oldStat.getStats().size() == 1) {
+          oldStat.replaceWith(caseStat);
```

Comment:
If oldStat isn't used here afterwards, I think it's better to let SequenceHelper deal with it instead- it should be called after switch pattern processing is done

Mainly because that is it's entire goal, to clean up weird stuff with sequences afterwards so that the actual pipeline passes don't need to deal with it (and because it does some other stuff, like eliminating empty basic blocks that can be inlined)